### PR TITLE
1D NumPy frontend

### LINKED ISSUE #497: axis argument torch.stack is undocumented
We should use `dim` instead, per https://pytorch.org/docs/stable/torch.html#torch.stack.

### PR DESCRIPTION
**Pull request by janden**
Closes #497 

The other PR (#499) was too messy due to rebases and history conflicts, so I cherry-picked the top three commits onto 0.2.X and made this PR instead. If anything's missing, @edouardoyallon, let me know.

### CODE REVIEW

File: kymatio/scattering1d/backend/numpy_backend.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+
+    Returns
+    -------
+    res : tensor
+        The input tensor periodized along the next to last axis to yield a
+        tensor of size x.shape[-2] // k along that dimension.
+    """
+
+    N = x.shape[-2]
+    res = x.view(x.shape[:-2] + (k, N // k )).mean(dim=-1)
```
Comment (janden) [inline kymatio/scattering1d/backend/numpy_backend.py]: Um... does `ndarray.view` work this way? Not according to [the docs](https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.view.html).

File: kymatio/scattering1d/backend/numpy_backend.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+    x : tensor
+        Three-dimensional input tensor with the third axis being the one to
+        be padded.
+    pad_left : int
+        Amount to add on the left of the tensor (at the beginning of the
+        temporal axis).
+    pad_right : int
+        amount to add on the right of the tensor (at the end of the temporal
+        axis).
+    to_complex : boolean, optional
```
Comment (janden) [inline kymatio/scattering1d/backend/numpy_backend.py]: Not used. Remove.

File: kymatio/scattering1d/backend/numpy_backend.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+
+    Returns
+    -------
+    output : tensor
+        A padded signal, possibly transformed into a four-dimensional tensor
+        with the last axis of size 2 if to_complex is True (this axis
+        corresponds to the real and imaginary parts).
+    """
+    output = pad_1d(x, pad_left, pad_right, mode='reflect')
+    #if to_complex:
```
Comment (janden) [inline kymatio/scattering1d/backend/numpy_backend.py]: Remove commented-out code.

File: kymatio/scattering1d/backend/numpy_backend.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+    x : tensor
+        A tensor of the same size of x_f containing the normalized inverse
+        Fourier transform of x_f.
+    """
+    return np.fft.ifft(x)
+
+def concatenate(arrays):
+    return np.stack(arrays, axis=-1)
+
+backend = namedtuple('backend', ['name', 'modulus_complex', 'subsample_fourier', 'real', 'unpad', 'fft1d_c2c',\
```
Comment (janden) [inline kymatio/scattering1d/backend/numpy_backend.py]: No need for backslash (I think).

File: kymatio/scattering1d/frontend/numpy_frontend.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+
+class ScatteringNumPy1D(ScatteringNumPy, ScatteringBase1D):
+    def __init__(self, J, shape, Q=1, max_order=2, average=True, oversampling=0, vectorize=True, backend='numpy'):
+        ScatteringNumPy.__init__(self)
+        ScatteringBase1D.__init__(self, J, shape, Q, max_order, average, oversampling, vectorize, backend)
+        ScatteringBase1D._instantiate_backend(self, 'kymatio.scattering1d.backend.')
+        ScatteringBase1D.build(self)
+        ScatteringBase1D.create_filters(self)
+
```
Comment (janden) [inline kymatio/scattering1d/frontend/numpy_frontend.py]: Remove extra newline.

File: kymatio/scattering1d/frontend/numpy_frontend.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+                raise ValueError(
+                    'Options average=False and vectorize=True are ' +
+                    'mutually incompatible. Please set vectorize to False.')
+            size_scattering = precompute_size_scattering(
+                self.J, self.Q, max_order=self.max_order, detail=True)
+        else:
+            size_scattering = 0
+
+        S = scattering1d(x, self.backend.pad, self.backend.unpad, self.backend, self.J, self.psi1_f, self.psi2_f,
+                         self.phi_f, \
```
Comment (janden) [inline kymatio/scattering1d/frontend/numpy_frontend.py]: Remove backslash.

File: kymatio/scattering1d/frontend/numpy_frontend.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+                    len(x.shape)))
+
+        batch_shape = x.shape[:-1]
+        signal_shape = x.shape[-1:]
+
+        x = x.reshape((-1, 1) + signal_shape)
+
+        # get the arguments before calling the scattering
+        # treat the arguments
+        if self.vectorize:
```
Comment (janden) [inline kymatio/scattering1d/frontend/numpy_frontend.py]: This stuff is generic and should probably be in the `ScatteringBase1D` class. Not necessary now, but we might want to make an issue.

File: kymatio/scattering1d/tests/test_numpy_scattering1d.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+backends = []
+
+from kymatio.scattering1d.backend.numpy_backend import backend
+backends.append(backend)
+
+
+
+class TestScattering1DNumpy:
+
```
Comment (janden) [inline kymatio/scattering1d/tests/test_numpy_scattering1d.py]: Remove extra blank line.

File: kymatio/scattering1d/tests/test_numpy_scattering1d.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+import numpy as np
+import torch
+
+backends = []
+
+from kymatio.scattering1d.backend.numpy_backend import backend
+backends.append(backend)
+
+
+
```
Comment (janden) [inline kymatio/scattering1d/tests/test_numpy_scattering1d.py]: Remove extra blank line.

File: kymatio/scattering1d/tests/test_numpy_scattering1d.py [hunk 10]
```diff
@@ -0,0 +1,67 @@
+import pytest
+from kymatio import Scattering1D
+import math
+import os
+import numpy as np
+import torch
```
Comment (janden) [inline kymatio/scattering1d/tests/test_numpy_scattering1d.py]: No.

File: kymatio/scattering1d/tests/test_numpy_scattering1d.py [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+class TestScattering1DNumpy:
+
+    @pytest.mark.parametrize('backend', backends)
+    def test_Scattering1D(self, backend):
+        """
+        Applies scattering on a stored signal to make sure its output agrees with
+        a previously calculated version.
+        """
+        test_data_dir = os.path.dirname(__file__)
+        test_data_filename = os.path.join(test_data_dir, 'test_data_1d.pt')
```
Comment (janden) [inline kymatio/scattering1d/tests/test_numpy_scattering1d.py]: Use the new data.

File: kymatio/scattering1d/tests/test_numpy_scattering1d.py [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+            for n2 in n2s:
+                orders[2].append(ind)
+                ind = ind + 1
+
+        perm = np.concatenate(orders)
+
+        Sx0 = Sx0[:, perm, :]
+
+        Sx = scattering(x)
+        assert np.allclose(Sx, Sx0)
```
Comment (janden) [inline kymatio/scattering1d/tests/test_numpy_scattering1d.py]: Add newline here.

File: kymatio/scattering1d/tests/test_numpy_scattering1d.py [hunk 13]
```diff
@@ -0,0 +1,67 @@
+import pytest
```
Comment (janden) [inline kymatio/scattering1d/tests/test_numpy_scattering1d.py]: These tests don't cover the whole set of functionality (like the torch tests and the 2D NP tests). This is fine, for now, but we should make an issue and fix this.

**Comment by edouardoyallon** [comment]
Right, so it's really too bad because I have a PR which passes the tests...

File: kymatio/scattering1d/tests/test_numpy_scattering1d.py (hunk 13 above)
Comment (edouardoyallon) [reply kymatio/scattering1d/tests/test_numpy_scattering1d.py]: I agree.

**Comment by janden** [comment]
> Right, so it's really too bad because I have a PR which passes the tests...

Passes the tests? It didn't look like it last time I checked. At any rate, if your only changes are the last three commits, it should be fine.

**Comment by edouardoyallon** [comment]
> > Right, so it's really too bad because I have a PR which passes the tests...
> 
> Passes the tests? It didn't look like it last time I checked. At any rate, if your only changes are the last three commits, it should be fine.

C'est bon, ça marche.

File: kymatio/scattering1d/tests/test_numpy_scattering1d.py (hunk 11 above)
Comment (edouardoyallon) [reply kymatio/scattering1d/tests/test_numpy_scattering1d.py]: This is now fixed

**Comment by janden** [comment]
> C'est bon, ça marche.

:+1:

Will go through and merge tomorrow.

File: kymatio/scattering1d/frontend/numpy_frontend.py (hunk 7 above)
Comment (edouardoyallon) [reply kymatio/scattering1d/frontend/numpy_frontend.py]: OK

**Comment by edouardoyallon** [comment]
comments addressed.

File: kymatio/scattering1d/backend/numpy_backend.py (hunk 2 above)
Comment (janden) [reply kymatio/scattering1d/backend/numpy_backend.py]: Still here. Remove.

File: kymatio/scattering1d/backend/numpy_backend.py [hunk 14]
```diff
@@ -119,15 +111,14 @@ def pad(x, pad_left=0, pad_right=0, to_complex=True):
         corresponds to the real and imaginary parts).
     """
     output = pad_1d(x, pad_left, pad_right, mode='reflect')
-    #if to_complex:
-        #output = torch.stack((output, torch.zeros_like(output)), dim=-1)
+    # if to_complex:
```
Comment (janden) [inline kymatio/scattering1d/backend/numpy_backend.py]: Remove commented-out code.

File: kymatio/scattering1d/backend/numpy_backend.py [hunk 15]
```diff
@@ -197,10 +183,12 @@ def ifft1d_c2c(x):
     """
     return np.fft.ifft(x)
 
+
 def concatenate(arrays):
     return np.stack(arrays, axis=-1)
 
-backend = namedtuple('backend', ['name', 'modulus_complex', 'subsample_fourier', 'real', 'unpad', 'fft1d_c2c',\
+
+backend = namedtuple('backend', ['name', 'modulus_complex', 'subsample_fourier', 'real', 'unpad', 'fft1d_c2c', \
```
Comment (janden) [inline kymatio/scattering1d/backend/numpy_backend.py]: No backslash.

File: kymatio/scattering1d/backend/numpy_backend.py [hunk 16]
```diff
@@ -211,4 +199,4 @@ def concatenate(arrays):
 backend.pad_1d = pad_1d
 backend.fft1d_c2c = fft1d_c2c
 backend.ifft1d_c2c = ifft1d_c2c
-backend.concatenate = concatenate
+backend.concatenate = concatenate
```
Comment (janden) [inline kymatio/scattering1d/backend/numpy_backend.py]: Add newline.

File: kymatio/scattering1d/tests/test_numpy_scattering1d.py (hunk 12 above)
Comment (janden) [reply kymatio/scattering1d/tests/test_numpy_scattering1d.py]: Still not here.
